pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1150 @@
+                let backoff = utils::backoff::exponential_backoff_duration_seconds(
+                    consecutive_failures.min(u32::MAX as usize) as u32,
+                    1.5,
+                    60.0,
+                );
+                let backoff = std::time::Duration::from_secs_f64(backoff);
+
+                // unless we get cancelled, we will hold off the init semaphore
+                tokio::time::sleep(backoff).await;
+
+                Err(DownloadError::DownloadFailed)
+            }
+            Err(_gone) => Err(DownloadError::DownloadCancelled),
+        }
+    }
+
+    async fn needs_download(&self) -> Result<Option<NeedsDownload>, std::io::Error> {
+        match tokio::fs::metadata(&self.path).await {
+            Ok(m) => Ok(self.is_file_present_and_good_size(&m).err()),
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(Some(NeedsDownload::NotFound)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn needs_download_blocking(&self) -> Result<Option<NeedsDownload>, std::io::Error> {
+        match self.path.metadata() {
+            Ok(m) => Ok(self.is_file_present_and_good_size(&m).err()),
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(Some(NeedsDownload::NotFound)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn is_file_present_and_good_size(&self, m: &std::fs::Metadata) -> Result<(), NeedsDownload> {
+        // in future, this should include sha2-256 of the file.
+        if !m.is_file() {
+            Err(NeedsDownload::NotFile)
+        } else if m.len() != self.desc.file_size {
+            Err(NeedsDownload::WrongSize {
+                actual: m.len(),
+                expected: self.desc.file_size,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    fn info(&self, reset: LayerAccessStatsReset) -> HistoricLayerInfo {
+        let layer_file_name = self.desc.filename().file_name();
+
+        let remote = self
+            .needs_download_blocking()
+            .map(|maybe| maybe.is_some())
+            .unwrap_or(true);
+
+        let access_stats = self.access_stats.as_api_model(reset);
+
+        if self.desc.is_delta {
+            let lsn_range = &self.desc.lsn_range;
+
+            HistoricLayerInfo::Delta {
+                layer_file_name,
+                layer_file_size: self.desc.file_size,
+                lsn_start: lsn_range.start,
+                lsn_end: lsn_range.end,
+                remote,
+                access_stats,
+            }
+        } else {
+            let lsn = self.desc.image_layer_lsn();
+
+            HistoricLayerInfo::Image {
+                layer_file_name,
+                layer_file_size: self.desc.file_size,
+                lsn_start: lsn,
+                remote,
+                access_stats,
+            }
+        }
+    }
+
+    /// Our resident layer has been dropped, we might hold the lock elsewhere.
+    fn on_drop(self: Arc<LayerInner>) {
```

Comment:
This is a method on `LayerInner`, and the name "on_drop" suggests that it's about "on drop of LayerInner".
But really, it's "on drop of the DownloadedLayer that references this LayerInner".

Also, no idea what the second half of the sentence is supposed to mean (", we might hold the lock elsewhere")

```suggestion
    /// Our [`DownloadedLayer`] has been dropped.
    fn on_drop_of_downloaded_layer(self: Arc<LayerInner>) {
```

Reply:
I was thinking we could be holding on to the heavier_once_cell::OnceCell::0 which is a plain mutex, but good spotting, should clean this up.

Reply:
b4c81f7df